### PR TITLE
Spring boot package [Memory - data stream]

### PR DESCRIPTION
**Pull request by sunny-elastic**
## What does this PR do?

- Generated the skeleton of Spring Boot integration package.
- Added **Memory** data stream
- Added data collection logic.
- Added the ingest pipelines.
- Mapped fields according to the ECS schema and added Fields metadata in the appropriate yml files.
- Added system test cases.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that Memory data stream collect metrics.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] If I'm introducing a new feature, I have modified the Kibana version constraint in my package's `manifest.yml` file to point to the latest Elastic stack release (e.g. `^8.0.0`).

## How to test this PR locally

- Clone integrations repo.
- Install elastic-package locally.
- Start elastic stack using elastic-package.
- Move to integrations/packages/spring_boot directory.
- Run the following command to run tests. 
> elastic-package test 

Note: We have covered dashboards and the visualisations for all data streams of spring boot into separate PR. Also Kibana version will be updated to 8.1.0 in manifest.yml after testing this integration on 8.1.0.

### CODE REVIEW

File: packages/spring_boot/data_stream/memory/sample_event.json [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        "address": "http://springboot:8090/actuator/jolokia",
+        "type": "jolokia"
+    },
+    "spring_boot": {
+        "memory": {
+            "buffer_pool": {
+                "direct": {
+                    "count": 10,
+                    "memory_used": 81920,
+                    "total_capacity": 81920
```
Comment (mtojek) [inline packages/spring_boot/data_stream/memory/sample_event.json:69]: Used and total are the same? Does it mean that you have 100% consumed the memory?

File: packages/spring_boot/data_stream/memory/sample_event.json [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            "codename": "focal",
+            "family": "debian",
+            "kernel": "3.10.0-1160.45.1.el7.x86_64",
+            "name": "Ubuntu",
+            "platform": "ubuntu",
+            "type": "linux",
+            "version": "20.04.3 LTS (Focal Fossa)"
+        }
+    },
+    "jolokia": {},
```
Comment (mtojek) [inline packages/spring_boot/data_stream/memory/sample_event.json:54]: I guess that you can strip empty Jolokia groups?

File: packages/spring_boot/data_stream/memory/sample_event.json [hunk 3]
```diff
@@ -0,0 +1,74 @@
+{
```
Comment (mtojek) [inline packages/spring_boot/data_stream/memory/sample_event.json:1]: Could you dump an event that collects most of defined metrics?

File: packages/spring_boot/data_stream/memory/fields/fields.yml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                  type: long
+                - name: init
+                  type: long
+                - name: max
+                  type: long
+                - name: used
+                  type: long
+            - name: object_pending_finalization_count
+              type: long
+            - name: verbose
```
Comment (mtojek) [inline packages/spring_boot/data_stream/memory/fields/fields.yml:55]: What does the verbose mean here?

File: packages/spring_boot/data_stream/memory/sample_event.json (hunk 2 above)
Comment (sunny-elastic) [reply packages/spring_boot/data_stream/memory/sample_event.json]: yes removed that

File: packages/spring_boot/data_stream/memory/fields/fields.yml [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      fields:
+        - name: buffer_pool
+          type: group
+          fields:
+            - name: direct
+              type: group
+              fields:
+                - name: count
+                  type: long
+                  description: Returns count of direct buffer pool memory
```
Comment (mtojek) [inline packages/spring_boot/data_stream/memory/fields/fields.yml:15]: Could you please make all descriptions consistent? Why is there everywhere a verb "Returns"? Let's make it consistent with other data streams.

File: packages/spring_boot/data_stream/memory/fields/fields.yml [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                  description: Returns used memory of direct buffer pool
+                - name: total_capacity
+                  type: long
+                  description: Returns total capacity of direct buffer pool memory
+            - name: mapped
+              type: group
+              fields:
+                - name: count
+                  type: long
+                  description: Returns count of mapped buffer pool memory
```
Comment (mtojek) [inline packages/spring_boot/data_stream/memory/fields/fields.yml:27]: What is this metric about?

File: packages/spring_boot/docs/README.md [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+| spring_boot.memory.memory.non_heap_memory_usage.committed | Returns committed non-heap memory usage of JVM | long |
+| spring_boot.memory.memory.non_heap_memory_usage.init | Returns init non-heap memory usage of JVM | long |
+| spring_boot.memory.memory.non_heap_memory_usage.max | Returns max non-heap memory usage of JVM | long |
+| spring_boot.memory.memory.non_heap_memory_usage.used | Returns used non-heap memory usage of JVM | long |
+| spring_boot.memory.memory.object_pending_finalization_count | Returns the approximate number of objects for which finalization is pending | long |
+| spring_boot.memory.memory.verbose | Tests if verbose output for the memory system is enabled | boolean |
+| spring_boot.memory.memory_manager.code_cache_manager.name | Name of the cacheManager to qualify the cache | keyword |
+| spring_boot.memory.memory_manager.code_cache_manager.valid | Return the validation | boolean |
+| spring_boot.memory.memory_manager.metaspace_manager.name | Name of the Metaspace Manager to qualify the cache | keyword |
+| spring_boot.memory.memory_manager.metaspace_manager.valid | Return the validation | boolean |
```
Comment (mtojek) [inline packages/spring_boot/docs/README.md:181]: Please work together with @yug-elastic to refactor those metrics. `memory` is repeated a few times per metric.

**Review by mtojek — CHANGES_REQUESTED** [review]

File: packages/spring_boot/docs/README.md (hunk 7 above)
Comment (yug-elastic) [reply packages/spring_boot/docs/README.md:181]: Sure, putting the link to the similar discussion for Apache Spark here for ready reference
https://github.com/elastic/integrations/pull/2811#discussion_r837187132

File: packages/spring_boot/data_stream/memory/sample_event.json (hunk 1 above)
Comment (sunny-elastic) [reply packages/spring_boot/data_stream/memory/sample_event.json]: Yes.

File: packages/spring_boot/data_stream/memory/fields/fields.yml (hunk 4 above)
Comment (sunny-elastic) [reply packages/spring_boot/data_stream/memory/fields/fields.yml]: Verbose: Tests if verbose output for the memory system is enabled. But I think its a boolean value and does not make any sense to collect that so we removed it.

File: packages/spring_boot/data_stream/memory/fields/fields.yml (hunk 6 above)
Comment (sunny-elastic) [reply packages/spring_boot/data_stream/memory/fields/fields.yml]: Count of buffer pool memory

File: packages/spring_boot/docs/README.md (hunk 7 above)
Comment (sunny-elastic) [reply packages/spring_boot/docs/README.md]: Its updated

File: packages/spring_boot/data_stream/memory/sample_event.json (hunk 3 above)
Comment (sunny-elastic) [reply packages/spring_boot/data_stream/memory/sample_event.json:1]: updated.

File: packages/spring_boot/data_stream/memory/fields/fields.yml (hunk 6 above)
Comment (mtojek) [reply packages/spring_boot/data_stream/memory/fields/fields.yml]: Still, what does it mean :) Is it the number of buffer pools, number of memory units, etc.?

File: packages/spring_boot/_dev/build/docs/README.md [hunk 8]
```diff
@@ -47,3 +47,15 @@ This is the `http_trace` data stream.
...
 
 {{fields "http_trace"}}
+
+## Metrics
+
+### Memory Metrics
+
+This is the `memory` data stream.
+
+- This data stream gives Memory information.
```
Comment (mtojek) [inline packages/spring_boot/_dev/build/docs/README.md:57]: nit: memory

File: packages/spring_boot/data_stream/memory/fields/ecs.yml [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  level: core
+  name: source.geo.location
+  type: geo_point
+- external: ecs
+  name: source.geo.region_iso_code
+- external: ecs
+  name: source.geo.region_name
+- external: ecs
+  name: source.ip
+- external: ecs
```
Comment (mtojek) [inline packages/spring_boot/data_stream/memory/fields/ecs.yml:75]: This data stream is about memory, so why do we need all these field definitions? Process, source, tls, HTTP. Did you add them on purpose?

**Review by mtojek — CHANGES_REQUESTED** [review]

File: packages/spring_boot/data_stream/memory/fields/fields.yml (hunk 6 above)
Comment (sunny-elastic) [reply packages/spring_boot/data_stream/memory/fields/fields.yml]: yes it is number of buffer pools

File: packages/spring_boot/data_stream/memory/fields/ecs.yml (hunk 9 above)
Comment (sunny-elastic) [reply packages/spring_boot/data_stream/memory/fields/ecs.yml]: yeah removed those extra ecs fields

File: packages/spring_boot/_dev/build/docs/README.md [hunk 10]
```diff
@@ -47,3 +47,15 @@ This is the `http_trace` data stream.
...
 
 {{fields "http_trace"}}
+
+## Metrics
+
+### Memory Metrics
+
+This is the `memory` data stream.
+
+- This data stream gives memory information.
```
Comment (mtojek) [inline packages/spring_boot/_dev/build/docs/README.md:57]: I don't think that this description explains what kind of data the data stream is handling. Please rephrase it to something meaningful (what kind of metrics, from where).

File: packages/spring_boot/docs/README.md [hunk 11]
```diff
@@ -225,3 +225,136 @@ An example event for `http_trace` looks as following:
...
+| message | For log events the message field contains the log message, optimized for viewing in a log viewer. For structured logs without an original message field, other fields can be concatenated to form a human-readable summary of the event. If multiple messages exist, they can be combined into one message. | match_only_text |
+| os.name | Operating system name, without the version. | keyword |
+| os.name.text | Multi-field of `os.name`. | match_only_text |
+| os.version | Operating system version as a raw string. | keyword |
```
Comment (mtojek) [inline packages/spring_boot/docs/README.md:339]: Sample event doesn't contain metrics for these entries. Does the integration collect these values?

File: packages/spring_boot/docs/README.md [hunk 12]
```diff
@@ -225,3 +225,136 @@ An example event for `http_trace` looks as following:
...
+| event.module | Name of the module this data is coming from. If your monitoring agent supports the concept of modules or plugins to process events of a given source (e.g. Apache logs), `event.module` should contain the name of this module. | keyword |
+| event.outcome | This is one of four ECS Categorization Fields, and indicates the lowest level in the ECS category hierarchy. `event.outcome` simply denotes whether the event represents a success or a failure from the perspective of the entity that produced the event. Note that when a single transaction is described in multiple events, each event may populate different values of `event.outcome`, according to their perspective. Also note that in the case of a compound event (a single event that contains multiple logical events), this field should be populated with the value that best captures the overall success or failure from the perspective of the event producer. Further note that not all events will have an associated outcome. For example, this field is generally not populated for metric events, events with `event.type:info`, or any events for which an outcome does not make logical sense. | keyword |
```
Comment (mtojek) [inline packages/spring_boot/docs/README.md:335]: Same

File: packages/spring_boot/docs/README.md [hunk 13]
```diff
@@ -225,3 +225,136 @@ An example event for `http_trace` looks as following:
...
+| message | For log events the message field contains the log message, optimized for viewing in a log viewer. For structured logs without an original message field, other fields can be concatenated to form a human-readable summary of the event. If multiple messages exist, they can be combined into one message. | match_only_text |
```
Comment (mtojek) [inline packages/spring_boot/docs/README.md:336]: Same question. It's a metric based data stream. Why is there a message?

**Review by mtojek — CHANGES_REQUESTED** [review]

File: packages/spring_boot/_dev/build/docs/README.md (hunk 10 above)
Comment (sunny-elastic) [reply packages/spring_boot/_dev/build/docs/README.md]: Ok updated

File: packages/spring_boot/docs/README.md (hunk 11 above)
Comment (sunny-elastic) [reply packages/spring_boot/docs/README.md]: missed it from Readme, but its updated now

File: packages/spring_boot/docs/README.md (hunk 12 above)
Comment (sunny-elastic) [reply packages/spring_boot/docs/README.md:335]: yes updated

File: packages/spring_boot/docs/README.md (hunk 13 above)
Comment (sunny-elastic) [reply packages/spring_boot/docs/README.md]: yeah its cleared

File: packages/spring_boot/docs/README.md (hunk 12 above)
Comment (mtojek) [reply packages/spring_boot/docs/README.md:335]: What did you update here? I can still see `event.outcome` and I can't see in the sample event. Could you please explain why is this field here?

File: packages/spring_boot/docs/README.md [hunk 14]
```diff
@@ -225,3 +225,129 @@ An example event for `http_trace` looks as following:
...
+| spring_boot.memory.buffer_pool.mapped.total_capacity | Total capacity of mapped buffer pool memory | long |
+| spring_boot.memory.buffer_pool.mapped.used | Used memory of mapped buffer pool | long |
+| spring_boot.memory.heap.committed | Committed heap memory usage of JVM | long |
+| spring_boot.memory.heap.init | Init heap memory usage of JVM | long |
+| spring_boot.memory.heap.max | Max heap memory usage of JVM | long |
+| spring_boot.memory.heap.used | Used heap memory usage of JVM | long |
+| spring_boot.memory.manager.code_cache.name | Name of the cacheManager to qualify the cache | keyword |
+| spring_boot.memory.manager.code_cache.valid | Returns the validation | boolean |
+| spring_boot.memory.manager.metaspace.name | Name of the Metaspace Manager to qualify the cache | keyword |
+| spring_boot.memory.manager.metaspace.valid | Returns the validation | boolean |
```
Comment (kush-elastic) [inline packages/spring_boot/docs/README.md:349]: can we rephrase this to something like
```suggestion
| spring_boot.memory.manager.code_cache.valid | Validation of code cache | boolean |
| spring_boot.memory.manager.metaspace.name | Name of the Metaspace Manager to qualify the cache | keyword |
| spring_boot.memory.manager.metaspace.valid | Validation of metaspace manager | boolean |
```

File: packages/spring_boot/manifest.yml [hunk 15]
```diff
@@ -40,5 +40,67 @@ policy_templates:
...
+          - name: ssl
+            type: yaml
+            title: SSL Configuration
+            description: i.e. certificate_authorities, supported_protocols, verification_mode etc.
+            multi: false
+            required: false
+            show_user: false
+            default: |-
+              default: |
+                #certificate_authorities:
```
Comment (kush-elastic) [inline packages/spring_boot/manifest.yml:84]: I think there's a mistake here, maybe remove this default here:
```suggestion
            default: |-
              #certificate_authorities:
```

**Review by kush-elastic — CHANGES_REQUESTED** [review]

File: packages/spring_boot/docs/README.md (hunk 14 above)
Comment (sunny-elastic) [reply packages/spring_boot/docs/README.md]: Yes sure. It updated

File: packages/spring_boot/manifest.yml (hunk 15 above)
Comment (sunny-elastic) [reply packages/spring_boot/manifest.yml]: fixed

**Review by kush-elastic — APPROVED** [review]
LGTM!

**Review by mtojek — APPROVED** [review]
Just a formality on my side. Thanks, @kush-elastic, for reviewing.
